lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -580,92 +690,648 @@ void crypto_bignum_clear(struct bignum *s)
 					size_t label_len __unused,
-					const uint8_t *src __unused,
-					size_t src_len __unused,
-					uint8_t *dst __unused,
-					size_t *dst_len __unused)
+					const uint8_t *src, size_t src_len,
+					uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int lmd_res, lmd_padding;
+	size_t blen;
+	size_t mod_size;
+	void *buf = NULL;
+	mbedtls_rsa_context rsa;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo = MBEDTLS_MD_NONE;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+
+	rsa.E = *(mbedtls_mpi *)key->e;
+	rsa.D = *(mbedtls_mpi *)key->d;
+	rsa.N = *(mbedtls_mpi *)key->n;
+	if (key->p && crypto_bignum_num_bytes(key->p)) {
+		rsa.P = *(mbedtls_mpi *)key->p;
+		rsa.Q = *(mbedtls_mpi *)key->q;
+		rsa.QP = *(mbedtls_mpi *)key->qp;
+		rsa.DP = *(mbedtls_mpi *)key->dp;
+		rsa.DQ = *(mbedtls_mpi *)key->dq;
+	}
+
+	/*
+	 * Use a temporary buffer since we don't know exactly how large
+	 * the required size of the out buffer without doing a partial
+	 * decrypt. We know the upper bound though.
+	 */
+	if (algo == TEE_ALG_RSAES_PKCS1_V1_5) {
+		mod_size = crypto_bignum_num_bytes(key->n);
+		blen = mod_size - 11;
+		lmd_padding = MBEDTLS_RSA_PKCS_V15;
+	} else {
+		/* Decoded message is always shorter than encrypted message */
+		blen = src_len;
+		lmd_padding = MBEDTLS_RSA_PKCS_V21;
+	}
+
+	rsa.len =  crypto_bignum_num_bytes(key->n);
+
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (pk_info == NULL) {
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto out;
+	}
+
+	/*
+	 * TEE_ALG_RSAES_PKCS1_V1_5 is invalid in hash. But its hash algo will
+	 * not be used in rsa, so skip it here.
+	 */
+	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
+		md_algo = tee_algo_to_mbedtls_hash_algo(algo);
+		if (md_algo == MBEDTLS_MD_NONE) {
+			res = TEE_ERROR_NOT_SUPPORTED;
+			goto out;
+		}
+	}
+
+	mbedtls_rsa_set_padding(&rsa, lmd_padding, md_algo);
+
+	mbedtls_ctr_drbg_init(&ctr_drbg);
+	lmd_res = mbedtls_ctr_drbg_seed(&ctr_drbg, mbd_rand, NULL, NULL, 0);
+	if (lmd_res != 0) {
+		FMSG("mbedtls_ctr_drbg_seed ret is 0x%x", -lmd_res);
+		res = TEE_ERROR_SECURITY;
+		goto out;
+	}
+
+	if (lmd_padding == MBEDTLS_RSA_PKCS_V15)
+		lmd_res = pk_info->decrypt_func(&rsa, src, src_len, buf, &blen,
+						blen, NULL, NULL);
+	else
+		lmd_res = pk_info->decrypt_func(&rsa, src, src_len, buf, &blen,
+						blen, mbedtls_ctr_drbg_random,
+						&ctr_drbg);
+	if (lmd_res != 0) {
+		FMSG("decrypt_func() returned 0x%x", -lmd_res);
+		res = get_tee_result(lmd_res);
+		goto out;
+	}
+
+	if (*dst_len < blen) {
+		*dst_len = blen;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto out;
+	}
+
+	res = TEE_SUCCESS;
+	*dst_len = blen;
+	memcpy(dst, buf, blen);
+out:
+	mbedtls_ctr_drbg_free(&ctr_drbg);
+	if (buf)
+		free(buf);
+	/* Reset mpi to skip freeing here, those mpis will be freed with key */
```

Comment:
These ~11 lines is a recurring pattern, how about adding a helper function:
```C
static void rsa_free(mbedtls_rsa_context *rsa);
```

Reply:
Same as above.